### PR TITLE
#554: Implement object-based permissions

### LINKED ISSUE #554: Implement multi-tenancy access controls: restrict users to specific areas of interest
I've read https://github.com/digitalocean/netbox/issues/446 and thought that it would be nice to have a more general and finer access control with an ability to restrict users (user groups) to view / edit / create objects only in specific Tenants, Tenant Groups, or Sites. And such access control should be enforced in UI as well as in API access.

That way one could give some limited access for the tenants' engineers themselves, or set some internal access boundaries (for the security-conscious).

### PR DESCRIPTION
**Pull request by jeremystretch**
### Closes #554: Implement object-based permissions

- Introduce the new users.ObjectPermission model for tracking object-based permissions assignment (replaces the built-in Permission model)
- Introduce RestrictedQuerySet to provide the `restrict()` method for filtering querysets based on assigned permissions
- Rewrite a lot of the existing views and tests to accommodate the new permissions framework

### CODE REVIEW

**Comment by mrfroggg** [comment]
This is huge! 👍

**Comment by itdependsnetworks** [comment]
I think https://github.com/netbox-community/netbox/blob/7b01ba9776fc06d91f0465e57b0c46efbb605542/netbox/users/models.py#L158
`@receiver(post_save, sender=User)` -> `@receiver(post_save, sender=AdminUser)`

File: netbox/users/admin.py [hunk 1]
```diff
@@ -16,14 +35,26 @@ class UserConfigInline(admin.TabularInline):
...
+@admin.register(AdminUser)
 class UserAdmin(UserAdmin_):
     list_display = [
         'username', 'email', 'first_name', 'last_name', 'is_superuser', 'is_staff', 'is_active'
     ]
+    fieldsets = (
+        (None, {'fields': ('username', 'password')}),
+        ('Personal info', {'fields': ('first_name', 'last_name', 'email')}),
+        ('Permissions', {
+            'fields': ('is_active', 'is_staff', 'is_superuser'),
```
Comment (itdependsnetworks) [inline netbox/users/admin.py:47]: I think groups need to be added.
`            'fields': ('is_active', 'is_staff', 'is_superuser', 'groups'),`

Comment (jeremystretch) [reply netbox/users/admin.py:47]: This is still a WIP and not ready for review.

**Pull request merged**